connaisseur/alert.py

```diff
@@ -0,0 +1,188 @@
+from connaisseur.exceptions import AlertSendingError, ConfigurationError
+from connaisseur.image import Image
+from connaisseur.mutate import get_container_specs
+
+
+class Alert:
+    """
+    Class to store image information about an alert as attributes and a sending functionality as method.
+    Alert Sending can, depending on the configuration, throw an AlertSendingError causing Connaisseur
+    responding with status code 500 to the request that was sent for admission control,
+    causing a Kubernetes Error event.
+    """
+
+    template: str
+    receiver_url: str
+    payload: dict
+    headers: dict
+
+    context: dict
+    admission_request: dict
+    throw_if_alert_sending_fails: bool
+
+    def __init__(self, alert_message, receiver_config, admission_request):
+        self.context = {
+            "alert_message": alert_message,
+            "priority": str(receiver_config.get("priority", 3)),
+            "connaisseur_pod_id": os.getenv("POD_NAME"),
+            "cluster": os.getenv("CLUSTER_NAME"),
+            "timestamp": datetime.now(),
+            "request_id": admission_request.get("request", {}).get(
+                "uid", "No given UID"
+            ),
+            "images": (str(get_images(admission_request)) or "No given images"),
+        }
+        self.admission_request = admission_request
+        self.receiver_url = receiver_config["receiver_url"]
+        self.template = receiver_config["template"]
+        self.throw_if_alert_sending_fails = receiver_config.get(
+            "fail_if_alert_sending_fails", False
+        )
+        self.payload = self._construct_payload(receiver_config)
+        self.headers = self._get_headers(receiver_config)
+
+    def _construct_payload(self, receiver_config):
+        try:
+            alert_templates_dir = f'{os.getenv("ALERT_CONFIG_DIR")}/templates'
+            with safe_path_func(
+                open,
+                alert_templates_dir,
+                f"{alert_templates_dir}/{self.template}.json",
+                "r",
+            ) as templatefile:
+                template = json.load(templatefile)
+        except Exception as err:
+            raise ConfigurationError(
+                "Template file for alerting payload is either missing or invalid JSON: {}".format(
+                    str(err)
+                )
+            ) from err
+        payload = self._render_template(template)
+        if receiver_config.get("payload_fields") is not None:
+            payload.update(receiver_config.get("payload_fields"))
+        return json.dumps(payload)
+
+    def _render_template(self, template):
+        if isinstance(template, dict):
+            for key in template.keys():
+                template[key] = self._render_template(template[key])
+        elif isinstance(template, list):
+            rendered_list = []
+            for entry in template:
+                rendered_list.append(self._render_template(entry))
+            template = rendered_list
```

Comment:
`template[:] = [self._render_template(entry) for entry in template]`

The slice `[:]` ensures that the original list is not modified during the loop, as I'm not sure whether that would have side effects on the iteration. It's possible that it can be dropped.

Reply:
Nice! Thx!